types.go

```diff
@@ -111,14 +119,19 @@ type Problem struct {
 	// Func evaluates the objective function at the given location. Func
 	// must not modify x.
 	Func func(x []float64) float64
+
```

Comment:
Not to change in this PR, but I wonder if we really want it to be the rule that x isn't modified, and instead just pass a copy of x. The copy should be reasonably cheap and then there are no bugs.